torch/csrc/jit/mobile/backport_factory.h

```diff
@@ -0,0 +1,79 @@
+#pragma once
+
+//#include <ATen/core/ivalue.h>
+#include <functional>
+#include <memory>
+#include <unordered_map>
+
+namespace c10 {
+struct IValue;
+}
+
+namespace caffe2 {
+namespace serialize {
+class IStreamAdapter;
+class ReadAdapterInterface;
+class PyTorchStreamWriter;
+class PyTorchStreamReader;
+} // namespace serialize
+} // namespace caffe2
+
+namespace torch {
+namespace jit {
+
+constexpr const char* kArchiveNameConstants = "constants";
+constexpr const char* kArchiveNameBytecode = "bytecode";
+constexpr const char* kArchiveNameVersion = "version";
+constexpr int64_t kBytecodeVersionV4 = 0x4L;
+constexpr int64_t kBytecodeVersionV5 = 0x5L;
+
+/*
+BackportFactory manages a list of backport from n to n-1 function, and provides
+function to check if a specific function exists.
+*/
+class BackportFactory final {
+ public:
+  bool hasBytecodeBackportFunction(const int64_t from_version) const;
+
+  std::unordered_map<
+      int64_t,
+      std::function<bool(
+          caffe2::serialize::PyTorchStreamReader&,
+          caffe2::serialize::PyTorchStreamWriter&)>>&
+  bytecodeBackportFunctions() const;
+
+  bool backport(
+      std::shared_ptr<caffe2::serialize::IStreamAdapter> istream_adapter,
+      caffe2::serialize::PyTorchStreamWriter& final_writer,
```

Comment:
@iseeyuan  I'm not very familiar with the serialization interfaces so just want to make sure this makes sense based on what we had discussed:

The reader, write was just me hacking something according the the types you were using, mostly to convey the higher level point.

So whether it's still a reader/writer or a single "mutable" interface that can read and write (not necessarily but potentially found in place changes) that could be nicer since to simplify the chaining semantics of the transformations.
Overall I don't feel strongly either way since this is a one time, one place usage :-)

Reply:
The process of the implementation includes conversions between stream and reader/writer.

Option 1, backport(reader, writer), the current implementation
```
for each version:
    inter_stream -> reader
    new output_stream->writer
    backport(reader, writer)
    if last version:
        writer->output_stream
    else:
        writer->inter_stream
```

Option 2, backport(stream):
```
def backport(stream):
    stream->reader
    backport_impl(reader->writer)
    writer->stream

for each version:
    backport(stream)
```

The difference is where to have the conversion logic. In option 1 it's in the loop which makes the loop look complicated. However it's one-time implementation. The developer who bumps the version has less code to write for the backport(reader, writer).

Option 2 moves the conversion logic inside backport function. However, if we implement a wrapper, the developer only need to write the backport_impl. The loop looks very clean.

I have no strong preference, just lean to option 2 for a cleaner pipe.